wurst/data/Buffer.wurst

```diff
@@ -0,0 +1,602 @@
+
+	Examples of serialized values:
+		1. "i50392|" - integer 50392
+		2. "r3.1415|" - real 3.1415
+		3. "b1" - boolean true
+		4. "s10|0123456789" - string 0123456789
+		5. "s5|01234i10|b1b0r1.0005|" - string 012345, integer 10, bool true, bool false, real 1.0005
+**/
+public class StringBuffer extends AbstractBuffer
+	private static constant TERMINATOR = "|"
+	private static constant MAX_BACKBUFFER_SIZE = 512
+	private static constant MAX_BUFFER_SIZE = 1024
+
+	// list of "chunks" that can grow/shrink as required
+	private var chunks = new LinkedList<ChunkElement>()
+	private var readBuffer = ""
+	private var writeBuffer = ""
+
+	// this is the maximum size of each chunk element
+	// we should keep this below 512 so that currentBuffer can reliably
+	// fit at least two buffers of that size, while staying under 1024 characters
+	private int maxBufferSize
+
+	construct(int maxSize)
+		if maxSize > MAX_BACKBUFFER_SIZE
+			Log.warn("Trying to instantiate an instance of StringBuffer with maxSize=" + maxSize.toString() +
+			", but MAX_BACKBUFFER_SIZE is " + MAX_BACKBUFFER_SIZE.toString())
+			Log.warn("The value will be clamped to " + MAX_BACKBUFFER_SIZE.toString())
+
+		maxBufferSize = min(maxSize, MAX_BACKBUFFER_SIZE)
+
+	/*
+		Write auxiliary methods.
+	*/
+
+	/**
+		Call this function once after you've finished writing to commit any pending data.
+	**/
+	function flush()
+		if writeBuffer.length() > 0
+			flushBuffer()
+
+	// flushes the current buffer to the chunks
+	private function flushBuffer()
+		chunks.push(new ChunkElement(writeBuffer))
+		writeBuffer = ""
+
+	// pushes a string to the current buffer, flushing if we exceed maxBufferSize
+	private function pushString(string data)
+		var remaining = data
+		if data.length() > MAX_BUFFER_SIZE
+			error("StringBuffer: trying to push a string above MAX_BUFFER_SIZE")
+
+		var overflow = data.length() + writeBuffer.length() - maxBufferSize
+		while overflow > 0
+			// we do not want the buffer to exceed maxBufferSize for each flush
+			let toAppend = min(remaining.length(), maxBufferSize - writeBuffer.length())
+			writeBuffer += remaining.substring(0, toAppend)
+			remaining = remaining.substring(toAppend)
+			overflow -= maxBufferSize
+			flushBuffer()
+
+		writeBuffer += remaining
+
+	private function pushTerminator()
+		pushString(TERMINATOR)
+
+	private function pushTypeIdentifier(ValueType valueType)
+		pushString(valueType.toString())
+
+	/*
+		Read auxiliary methods.
+	*/
+
+	private function isDataAvailable() returns boolean
+		return chunks.size() > 0 or readBuffer.length() > 0
+
+	private function checkDataAvailable()
+		if not isDataAvailable()
+			fail(BufferFailureMode.EOF, "reached EOF")
+
+	private function getNextChunkSize() returns integer
+		return chunks.getFirst().content.length()
+
+	// loads the next chunk into the buffer
+	private function nextChunk()
+		if chunks.size() == 0
+			fail(BufferFailureMode.EOF, "reached EOF")
+			return
+
+		let chunk = chunks.dequeue()
```

Comment:
using push and then dequeue kinda doesn't make sense. Please stick to sensible naming conventions of the intended datatype.

Reply:
should I use .add()/.dequeue() pair instead?